// @platejs/caption@53.0.0/dist/BaseCaptionPlugin-jZ8QfZJZ.js
import { KEYS, NodeApi, RangeApi, createTSlatePlugin, getPluginTypes, isHotkey } from "platejs";

//#region src/lib/withCaption.ts
/** TODO: tests https://github.com/udecode/editor-protocol/issues/79 */
/**
* Selection table:
*
* - If anchor is in table, focus in a block before: set focus to start of table
* - If anchor is in table, focus in a block after: set focus to end of table
* - If focus is in table, anchor in a block before: set focus to end of table
* - If focus is in table, anchor in a block after: set focus to the point before
*   start of table
*/
const withCaption = ({ editor, getOptions, tf: { apply, moveLine } }) => {
	return { transforms: {
		apply(operation) {
			const { query } = getOptions();
			if (operation.type === "set_selection") {
				const newSelection = {
					...editor.selection,
					...operation.newProperties
				};
				if (editor.dom.currentKeyboardEvent && isHotkey("up", editor.dom.currentKeyboardEvent) && newSelection && RangeApi.isCollapsed(newSelection)) {
					const types = getPluginTypes(editor, query.allow);
					const entry = editor.api.above({
						at: newSelection,
						match: { type: types }
					});
					if (entry) {
						const [node] = entry;
						if (node.caption && NodeApi.string({ children: node.caption }).length > 0) setTimeout(() => {
							editor.setOption(BaseCaptionPlugin, "focusEndPath", entry[1]);
						}, 0);
					}
				}
			}
			apply(operation);
		},
		moveLine: (options) => {
			const apply$1 = () => {
				if (!options.reverse) {
					const types = getPluginTypes(editor, getOptions().query.allow);
					const entry = editor.api.block({ match: { type: types } });
					if (!entry) return;
					editor.setOption(BaseCaptionPlugin, "focusEndPath", entry[1]);
					return true;
				}
			};
			if (apply$1()) return true;
			return moveLine(options);
		}
	} };
};

//#endregion
//#region src/lib/BaseCaptionPlugin.ts
/** Enables support for caption. */
const BaseCaptionPlugin = createTSlatePlugin({
	key: KEYS.caption,
	options: {
		focusEndPath: null,
		focusStartPath: null,
		query: { allow: [] },
		visibleId: null
	}
}).extendSelectors(({ getOptions }) => ({ isVisible: (elementId) => getOptions().visibleId === elementId })).overrideEditor(withCaption);

//#endregion
export { withCaption as n, BaseCaptionPlugin as t };
