// @platejs/link@53.3.1/dist/index.js
import { a as insertLink, c as validateUrl, d as encodeUrlIfNeeded, f as createLinkNode, i as unwrapLink, l as safeDecodeUrl, n as wrapLink, o as BaseLinkPlugin, r as upsertLinkText, s as withLink, t as upsertLink, u as getLinkAttributes } from "./upsertLink-DjO4x8_a.js";
import { KEYS, PathApi, createRuleFactory } from "platejs";

//#region src/lib/LinkRules.ts
const LINK_AUTOMD_REGEX = /\[([^\]\n]+)]\((\S+)$/;
const MARKDOWN_LINK_SOURCE_RE = /!?\[[^\]\n]*]\([^)\n]*$/;
const shouldAutoLinkPasteByDefault = (editor, { textBefore }) => {
	if (editor.api.some({ match: { type: [editor.getType(KEYS.codeBlock)] } })) return false;
	if (!editor.api.isCollapsed()) return true;
	return !MARKDOWN_LINK_SOURCE_RE.test(textBefore);
};
const getLinkAutomdMatch = (editor) => {
	const { selection } = editor;
	if (!selection || !editor.api.isCollapsed()) return;
	if (editor.api.some({ match: { type: [editor.getType(KEYS.codeBlock), editor.getType(KEYS.link)] } })) return;
	const blockRange = editor.api.range("start", selection);
	if (!blockRange) return;
	const textBefore = editor.api.string(blockRange);
	const match = LINK_AUTOMD_REGEX.exec(textBefore);
	if (!match) return;
	const [, linkText, rawUrl] = match;
	const { transformInput } = editor.getOptions(BaseLinkPlugin);
	const url = transformInput ? transformInput(rawUrl) ?? "" : rawUrl;
	if (!url || !validateUrl(editor, url)) return;
	const startPoint = editor.api.before(selection, {
		distance: match[0].length,
		unit: "character"
	});
	if (!startPoint) return;
	return {
		range: {
			anchor: startPoint,
			focus: selection.anchor
		},
		text: linkText,
		url
	};
};
const getAutolinkMatch = (editor) => {
	const { getUrlHref, isUrl: isUrl$1, rangeBeforeOptions } = editor.getOptions(BaseLinkPlugin);
	const { selection } = editor;
	if (!selection || !editor.api.isCollapsed()) return;
	let beforeWordRange = editor.api.range("before", selection, { before: rangeBeforeOptions });
	if (!beforeWordRange) beforeWordRange = editor.api.range("start", selection);
	if (!beforeWordRange) return;
	if (editor.api.some({
		at: beforeWordRange,
		match: { type: editor.getType(KEYS.link) }
	})) return;
	let beforeWordText = editor.api.string(beforeWordRange);
	beforeWordText = getUrlHref?.(beforeWordText) ?? beforeWordText;
	if (!isUrl$1?.(beforeWordText)) return;
	return {
		range: beforeWordRange,
		url: beforeWordText
	};
};
const applyAutolinkMatch = (editor, match) => {
	editor.tf.select(match.range);
	if (!upsertLink(editor, { url: match.url })) return false;
	editor.tf.collapse({ edge: "end" });
	return true;
};
const moveSelectionAfterLink = (editor) => {
	const { selection } = editor;
	if (!selection || !editor.api.isCollapsed()) return;
	const linkEntry = editor.api.above({ match: { type: editor.getType(KEYS.link) } });
	if (!linkEntry) return;
	const [, linkPath] = linkEntry;
	if (!editor.api.isEnd(selection.focus, linkPath)) return;
	const nextPoint = editor.api.start(linkPath, { next: true });
	if (nextPoint) {
		editor.tf.select(nextPoint);
		return;
	}
	const nextPath = PathApi.next(linkPath);
	editor.tf.insertNodes({ text: "" }, { at: nextPath });
	editor.tf.select(nextPath);
};
const LinkRules = {
	markdown: createRuleFactory({
		type: "insertText",
		trigger: ")",
		resolve: (context) => {
			if (context.text !== ")" || context.options?.at) return;
			return getLinkAutomdMatch(context.editor);
		},
		apply: (context, match) => {
			if (upsertLink(context.editor, {
				insertNodesOptions: {
					at: match.range,
					select: true
				},
				skipValidation: true,
				text: match.text,
				url: match.url
			})) {
				moveSelectionAfterLink(context.editor);
				return true;
			}
			context.insertText(context.text, context.options);
			return true;
		}
	}),
	autolink: createRuleFactory((options) => options.variant === "break" ? {
		type: "insertBreak",
		resolve: ({ editor }) => getAutolinkMatch(editor),
		apply: ({ editor }, match) => {
			if (!applyAutolinkMatch(editor, match)) return;
			moveSelectionAfterLink(editor);
			editor.tf.insertBreak();
			return true;
		}
	} : options.variant === "paste" ? {
		type: "insertData",
		resolve: (context) => {
			if (!context.text) return;
			const { getUrlHref } = context.editor.getOptions(BaseLinkPlugin);
			const url = getUrlHref?.(context.text) ?? context.text;
			if (!validateUrl(context.editor, url)) return;
			return {
				shouldLink: shouldAutoLinkPasteByDefault(context.editor, { textBefore: context.getBlockTextBeforeSelection() }),
				text: context.text,
				url
			};
		},
		apply: (context, match) => {
			const autolinkMatch = match;
			if (autolinkMatch.shouldLink) {
				const { keepSelectedTextOnPaste } = context.editor.getOptions(BaseLinkPlugin);
				if (upsertLink(context.editor, {
					insertTextInLink: true,
					text: keepSelectedTextOnPaste ? void 0 : autolinkMatch.url,
					url: autolinkMatch.url
				})) return true;
			}
			context.editor.tf.insertText(autolinkMatch.text);
			return true;
		}
	} : {
		type: "insertText",
		trigger: " ",
		resolve: (context) => {
			if (context.text !== " ") return;
			return getAutolinkMatch(context.editor);
		},
		apply: (context, match) => {
			const autolinkMatch = match;
			if (!applyAutolinkMatch(context.editor, autolinkMatch)) return;
			moveSelectionAfterLink(context.editor);
			context.insertText(context.text, context.options);
			return true;
		}
	})
};

//#endregion
export { BaseLinkPlugin, LinkRules, createLinkNode, encodeUrlIfNeeded, getLinkAttributes, insertLink, safeDecodeUrl, unwrapLink, upsertLink, upsertLinkText, validateUrl, withLink, wrapLink };
