// @platejs/math@53.0.0/dist/BaseInlineEquationPlugin-1MXNHk1G.js
import { KEYS, bindFirst, createSlatePlugin } from "platejs";
import "katex/dist/katex.min.css";

//#region src/lib/transforms/insertEquation.ts
const insertEquation = (editor, options) => {
	editor.tf.insertNodes({
		children: [{ text: "" }],
		texExpression: "",
		type: editor.getType(KEYS.equation)
	}, options);
};

//#endregion
//#region src/lib/transforms/insertInlineEquation.ts
const insertInlineEquation = (editor, texExpression, options) => {
	editor.tf.insertNodes({
		children: [{ text: "" }],
		texExpression: texExpression ?? editor.api.string(editor.selection),
		type: editor.getType(KEYS.inlineEquation)
	}, options);
};

//#endregion
//#region src/lib/BaseEquationPlugin.ts
const BaseEquationPlugin = createSlatePlugin({
	key: KEYS.equation,
	node: {
		isElement: true,
		isVoid: true
	}
}).extendEditorTransforms(({ editor }) => ({ insert: { equation: bindFirst(insertEquation, editor) } }));

//#endregion
//#region src/lib/BaseInlineEquationPlugin.ts
const BaseInlineEquationPlugin = createSlatePlugin({
	key: KEYS.inlineEquation,
	node: {
		isElement: true,
		isInline: true,
		isVoid: true
	}
}).extendEditorTransforms(({ editor }) => ({ insert: { inlineEquation: bindFirst(insertInlineEquation, editor) } }));

//#endregion
export { insertEquation as i, BaseEquationPlugin as n, insertInlineEquation as r, BaseInlineEquationPlugin as t };
