// @platejs/math@53.0.0/dist/index.js
import { i as insertEquation, n as BaseEquationPlugin, r as insertInlineEquation, t as BaseInlineEquationPlugin } from "./BaseInlineEquationPlugin-1MXNHk1G.js";
import { KEYS, createRuleFactory, matchDelimitedInline } from "platejs";
import katex from "katex";

//#region src/lib/MathRules.ts
const INLINE_BOUNDARY_RE = /[\s([{'"`]/;
const INLINE_FOLLOW_RE = /[\s)\]}:;,.!?'"`]/;
const isEquationInputBlocked = (editor) => editor.api.some({ match: { type: [
	editor.getType(KEYS.codeBlock),
	editor.getType(KEYS.equation),
	editor.getType(KEYS.inlineEquation)
] } });
const getInlineEquationMatch = (context) => {
	const match = matchDelimitedInline(context, {
		boundaryRe: INLINE_BOUNDARY_RE,
		followRe: INLINE_FOLLOW_RE,
		open: "$",
		requireClosingDelimiter: false,
		trim: "reject"
	});
	if (!match) return;
	return {
		deleteRange: match.deleteRange,
		texExpression: match.content
	};
};
const MathRules = { markdown: createRuleFactory((options) => options.variant === "$$" ? {
	type: "blockFence",
	fence: "$$",
	block: KEYS.p,
	on: options.on,
	enabled: ({ editor }) => !isEquationInputBlocked(editor),
	priority: 100,
	apply: ({ editor }, match) => {
		const blockMatch = match;
		editor.tf.removeNodes({ at: blockMatch.path });
		insertEquation(editor, {
			at: blockMatch.path,
			select: true
		});
		return true;
	}
} : {
	type: "insertText",
	enabled: ({ editor }) => !isEquationInputBlocked(editor),
	trigger: "$",
	resolve: (context) => {
		if (context.text !== "$" || context.options?.at) return;
		return getInlineEquationMatch(context);
	},
	apply: ({ editor }, match) => {
		const inlineMatch = match;
		editor.tf.delete({ at: inlineMatch.deleteRange });
		editor.tf.select(inlineMatch.deleteRange.anchor);
		insertInlineEquation(editor, inlineMatch.texExpression);
		return true;
	}
}) };

//#endregion
//#region src/lib/utils/getEquationHtml.ts
const getEquationHtml = ({ element, options }) => katex.renderToString(element.texExpression, options);

//#endregion
export { BaseEquationPlugin, BaseInlineEquationPlugin, MathRules, getEquationHtml, insertEquation, insertInlineEquation };
