// @platejs/caption@53.0.0/dist/react/index.js
import { t as BaseCaptionPlugin } from "../BaseCaptionPlugin-jZ8QfZJZ.js";
import { NodeApi, PathApi, isHotkey } from "platejs";
import { createPrimitiveComponent, toPlatePlugin, useEditorRef, useElement, useIsomorphicLayoutEffect, usePluginOption, useReadOnly, useSelected } from "platejs/react";
import { c } from "react-compiler-runtime";
import React, { useState } from "react";
import ReactTextareaAutosize from "react-textarea-autosize";

//#region src/react/CaptionPlugin.tsx
const CaptionPlugin = toPlatePlugin(BaseCaptionPlugin);

//#endregion
//#region src/react/hooks/useCaptionString.ts
const useCaptionString = () => {
	const $ = c(4);
	const { caption: t0 } = useElement();
	let t1;
	if ($[0] !== t0) {
		t1 = t0 === void 0 ? [{ children: [{ text: "" }] }] : t0;
		$[0] = t0;
		$[1] = t1;
	} else t1 = $[1];
	const nodeCaption = t1;
	let t2;
	if ($[2] !== nodeCaption[0]) {
		t2 = NodeApi.string(nodeCaption[0]) || "";
		$[2] = nodeCaption[0];
		$[3] = t2;
	} else t2 = $[3];
	return t2;
};

//#endregion
//#region src/react/components/Caption.tsx
const useCaptionState = (t0) => {
	const $ = c(7);
	let t1;
	if ($[0] !== t0) {
		t1 = t0 === void 0 ? {} : t0;
		$[0] = t0;
		$[1] = t1;
	} else t1 = $[1];
	const options = t1;
	const element = useElement();
	const captionString = useCaptionString();
	const showCaption$1 = usePluginOption(CaptionPlugin, "isVisible", element.id);
	const selected = useSelected();
	const _readOnly = useReadOnly();
	const readOnly = options.readOnly || _readOnly;
	const hidden = !showCaption$1 && captionString.length === 0;
	let t2;
	if ($[2] !== captionString || $[3] !== hidden || $[4] !== readOnly || $[5] !== selected) {
		t2 = {
			captionString,
			hidden,
			readOnly,
			selected
		};
		$[2] = captionString;
		$[3] = hidden;
		$[4] = readOnly;
		$[5] = selected;
		$[6] = t2;
	} else t2 = $[6];
	return t2;
};
const useCaption = (state) => ({ hidden: state.hidden });
const Caption = createPrimitiveComponent("figcaption")({
	propsHook: useCaption,
	stateHook: useCaptionState
});

//#endregion
//#region src/react/components/CaptionButton.tsx
const useCaptionButtonState = () => {
	const $ = c(3);
	const editor = useEditorRef();
	const element = useElement();
	let t0;
	if ($[0] !== editor || $[1] !== element) {
		t0 = {
			editor,
			element
		};
		$[0] = editor;
		$[1] = element;
		$[2] = t0;
	} else t0 = $[2];
	return t0;
};
const useCaptionButton = ({ editor, element }) => ({ props: { onClick: () => {
	const path = editor.api.findPath(element);
	editor.setOption(BaseCaptionPlugin, "visibleId", element.id);
	setTimeout(() => {
		if (path) editor.setOption(BaseCaptionPlugin, "focusEndPath", path);
	}, 0);
} } });

//#endregion
//#region src/react/components/TextareaAutosize.tsx
/**
* `<textarea />` component for React which grows with content.
*
* @see https://github.com/Andarist/react-textarea-autosize
* @see https://github.com/Andarist/react-textarea-autosize/issues/337
*/
const TextareaAutosize = React.forwardRef((props, ref) => {
	const $ = c(6);
	const [isRerendered, setIsRerendered] = React.useState(false);
	let t0;
	let t1;
	if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
		t0 = () => setIsRerendered(true);
		t1 = [];
		$[0] = t0;
		$[1] = t1;
	} else {
		t0 = $[0];
		t1 = $[1];
	}
	useIsomorphicLayoutEffect(t0, t1);
	let t2;
	if ($[2] !== isRerendered || $[3] !== props || $[4] !== ref) {
		t2 = isRerendered ? /* @__PURE__ */ React.createElement(ReactTextareaAutosize, {
			...props,
			ref
		}) : null;
		$[2] = isRerendered;
		$[3] = props;
		$[4] = ref;
		$[5] = t2;
	} else t2 = $[5];
	return t2;
});
TextareaAutosize.displayName = "TextareaAutosize";

//#endregion
//#region src/react/components/CaptionTextarea.tsx
/** Focus textareaRef when focusCaptionPath is set to the image path. */
const useCaptionTextareaFocus = (textareaRef) => {
	const $ = c(6);
	const editor = useEditorRef();
	const element = useElement();
	const focusCaptionPath = usePluginOption(CaptionPlugin, "focusEndPath");
	let t0;
	let t1;
	if ($[0] !== editor || $[1] !== element || $[2] !== focusCaptionPath || $[3] !== textareaRef) {
		t0 = () => {
			if (focusCaptionPath && textareaRef.current) {
				const path = editor.api.findPath(element);
				if (path && PathApi.equals(path, focusCaptionPath)) {
					textareaRef.current.focus();
					editor.setOption(CaptionPlugin, "focusEndPath", null);
				}
			}
		};
		t1 = [
			editor,
			element,
			focusCaptionPath,
			textareaRef
		];
		$[0] = editor;
		$[1] = element;
		$[2] = focusCaptionPath;
		$[3] = textareaRef;
		$[4] = t0;
		$[5] = t1;
	} else {
		t0 = $[4];
		t1 = $[5];
	}
	React.useEffect(t0, t1);
};
const useCaptionTextareaState = () => {
	const $ = c(17);
	const element = useElement();
	const editor = useEditorRef();
	const [isComposing, setIsComposing] = useState(false);
	let t0;
	if ($[0] !== element.caption) {
		t0 = () => {
			const nodeCaption = element.caption ?? [{ children: [{ text: "" }] }];
			return NodeApi.string(nodeCaption[0]);
		};
		$[0] = element.caption;
		$[1] = t0;
	} else t0 = $[1];
	const [captionValue, setCaptionValue] = useState(t0);
	let t1;
	if ($[2] !== editor || $[3] !== element) {
		t1 = (newValue) => {
			editor.tf.setNodes({ caption: [{ text: newValue }] }, { at: element });
		};
		$[2] = editor;
		$[3] = element;
		$[4] = t1;
	} else t1 = $[4];
	const updateEditorCaptionValue = t1;
	let t2;
	if ($[5] !== isComposing || $[6] !== updateEditorCaptionValue) {
		t2 = (e) => {
			const newValue_0 = e.target.value;
			setCaptionValue(newValue_0);
			if (!isComposing) updateEditorCaptionValue(newValue_0);
		};
		$[5] = isComposing;
		$[6] = updateEditorCaptionValue;
		$[7] = t2;
	} else t2 = $[7];
	const handleChange = t2;
	let t3;
	if ($[8] === Symbol.for("react.memo_cache_sentinel")) {
		t3 = () => {
			setIsComposing(true);
		};
		$[8] = t3;
	} else t3 = $[8];
	const handleCompositionStart = t3;
	let t4;
	if ($[9] !== updateEditorCaptionValue) {
		t4 = (e_0) => {
			setIsComposing(false);
			const newValue_1 = e_0.currentTarget.value;
			setCaptionValue(newValue_1);
			updateEditorCaptionValue(newValue_1);
		};
		$[9] = updateEditorCaptionValue;
		$[10] = t4;
	} else t4 = $[10];
	const handleCompositionEnd = t4;
	const readOnly = useReadOnly();
	const textareaRef = React.useRef(null);
	useCaptionTextareaFocus(textareaRef);
	let t5;
	if ($[11] !== captionValue || $[12] !== element || $[13] !== handleChange || $[14] !== handleCompositionEnd || $[15] !== readOnly) {
		t5 = {
			captionValue,
			element,
			readOnly,
			textareaRef,
			handleChange,
			handleCompositionEnd,
			handleCompositionStart
		};
		$[11] = captionValue;
		$[12] = element;
		$[13] = handleChange;
		$[14] = handleCompositionEnd;
		$[15] = readOnly;
		$[16] = t5;
	} else t5 = $[16];
	return t5;
};
const useCaptionTextarea = (t0) => {
	const $ = c(16);
	const { captionValue, element, readOnly, textareaRef, handleChange, handleCompositionEnd, handleCompositionStart } = t0;
	const editor = useEditorRef();
	let t1;
	if ($[0] !== editor || $[1] !== element) {
		t1 = (e) => {
			if (isHotkey("up", e)) {
				const path = editor.api.findPath(element);
				if (!path) return;
				e.preventDefault();
				editor.tf.focus({ at: path });
			}
			if (isHotkey("down", e)) {
				const path_0 = editor.api.findPath(element);
				if (!path_0) return;
				const nextNodePath = editor.api.after(path_0);
				if (!nextNodePath) return;
				e.preventDefault();
				editor.tf.focus({ at: nextNodePath });
			}
		};
		$[0] = editor;
		$[1] = element;
		$[2] = t1;
	} else t1 = $[2];
	const onKeyDown = t1;
	let t2;
	if ($[3] !== editor) {
		t2 = (e_0) => {
			if (e_0.target.value.length === 0) editor.setOption(CaptionPlugin, "visibleId", null);
		};
		$[3] = editor;
		$[4] = t2;
	} else t2 = $[4];
	const onBlur = t2;
	let t3;
	if ($[5] !== captionValue || $[6] !== handleChange || $[7] !== handleCompositionEnd || $[8] !== handleCompositionStart || $[9] !== onBlur || $[10] !== onKeyDown || $[11] !== readOnly) {
		t3 = {
			readOnly,
			value: captionValue,
			onBlur,
			onChange: handleChange,
			onCompositionEnd: handleCompositionEnd,
			onCompositionStart: handleCompositionStart,
			onKeyDown
		};
		$[5] = captionValue;
		$[6] = handleChange;
		$[7] = handleCompositionEnd;
		$[8] = handleCompositionStart;
		$[9] = onBlur;
		$[10] = onKeyDown;
		$[11] = readOnly;
		$[12] = t3;
	} else t3 = $[12];
	let t4;
	if ($[13] !== t3 || $[14] !== textareaRef) {
		t4 = {
			props: t3,
			ref: textareaRef
		};
		$[13] = t3;
		$[14] = textareaRef;
		$[15] = t4;
	} else t4 = $[15];
	return t4;
};
const CaptionTextarea = createPrimitiveComponent(TextareaAutosize)({
	propsHook: useCaptionTextarea,
	stateHook: useCaptionTextareaState
});

//#endregion
//#region src/react/utils/showCaption.ts
const showCaption = (editor, element) => {
	const path = editor.api.findPath(element);
	editor.setOption(CaptionPlugin, "visibleId", element.id);
	setTimeout(() => {
		if (path) editor.setOption(CaptionPlugin, "focusEndPath", path);
	}, 0);
};

//#endregion
export { Caption, CaptionPlugin, CaptionTextarea, TextareaAutosize, showCaption, useCaption, useCaptionButton, useCaptionButtonState, useCaptionState, useCaptionString, useCaptionTextarea, useCaptionTextareaFocus, useCaptionTextareaState };
