// @platejs/layout@53.0.0/dist/BaseColumnPlugin-z1GRkW4o.js
import { ElementApi, KEYS, PathApi, createSlatePlugin } from "platejs";

//#region src/lib/withColumn.ts
const withColumn = ({ editor, tf: { normalizeNode, selectAll }, type }) => ({ transforms: {
	normalizeNode([n, path]) {
		if (ElementApi.isElement(n) && n.type === editor.getType(KEYS.columnGroup)) {
			const node = n;
			const firstChild = node.children[0];
			if (node.children.length === 1 && firstChild.type === editor.getType(KEYS.p)) {
				editor.tf.unwrapNodes({ at: path });
				return;
			}
			if (!node.children.some((child) => ElementApi.isElement(child) && child.type === type)) {
				editor.tf.unwrapNodes({ at: path });
				return;
			}
			if (node.children.length < 2) {
				editor.tf.withoutNormalizing(() => {
					editor.tf.unwrapNodes({ at: path });
					editor.tf.unwrapNodes({ at: path });
				});
				return;
			}
			editor.tf.withoutNormalizing(() => {
				const totalColumns = node.children.length;
				let widths = node.children.map((col) => {
					const parsed = Number.parseFloat(col.width);
					return Number.isNaN(parsed) ? 0 : parsed;
				});
				const sum = widths.reduce((acc, w) => acc + w, 0);
				if (sum !== 100) {
					const adjustment = (100 - sum) / totalColumns;
					widths = widths.map((w) => w + adjustment);
					widths.forEach((w, i) => {
						const columnPath = path.concat([i]);
						editor.tf.setNodes({ width: `${w}%` }, { at: columnPath });
					});
				}
			});
		}
		if (ElementApi.isElement(n) && n.type === type) {
			if (n.children.length === 0) {
				editor.tf.removeNodes({ at: path });
				return;
			}
		}
		return normalizeNode([n, path]);
	},
	selectAll: () => {
		const apply = () => {
			const at = editor.selection;
			if (!at) return;
			const column = editor.api.above({ match: { type } });
			if (!column) return;
			let targetPath = column[1];
			if (editor.api.isStart(editor.api.start(at), targetPath) && editor.api.isEnd(editor.api.end(at), targetPath)) targetPath = PathApi.parent(targetPath);
			if (targetPath.length === 0) return;
			editor.tf.select(targetPath);
			return true;
		};
		if (apply()) return true;
		return selectAll();
	}
} });

//#endregion
//#region src/lib/BaseColumnPlugin.ts
const BaseColumnItemPlugin = createSlatePlugin({
	key: KEYS.column,
	node: {
		isContainer: true,
		isElement: true,
		isStrictSiblings: true
	}
}).overrideEditor(withColumn);
const BaseColumnPlugin = createSlatePlugin({
	key: KEYS.columnGroup,
	node: {
		isContainer: true,
		isElement: true
	},
	plugins: [BaseColumnItemPlugin]
});

//#endregion
export { BaseColumnPlugin as n, withColumn as r, BaseColumnItemPlugin as t };
