// @platejs/layout@53.0.0/dist/index.js
import { n as BaseColumnPlugin, r as withColumn, t as BaseColumnItemPlugin } from "./BaseColumnPlugin-z1GRkW4o.js";
import { KEYS, NodeApi } from "platejs";

//#region src/lib/transforms/insertColumn.ts
const insertColumn = (editor, { width = "33%", ...options } = {}) => {
	editor.tf.insertNodes({
		children: [editor.api.create.block()],
		type: editor.getType(KEYS.column),
		width
	}, options);
};

//#endregion
//#region src/lib/transforms/insertColumnGroup.ts
const insertColumnGroup = (editor, { columns = 2, select: selectProp, ...options } = {}) => {
	const width = 100 / columns;
	editor.tf.withoutNormalizing(() => {
		editor.tf.insertNodes({
			children: new Array(columns).fill(null).map(() => ({
				children: [editor.api.create.block()],
				type: editor.getType(KEYS.column),
				width: `${width}%`
			})),
			type: editor.getType(KEYS.columnGroup)
		}, options);
		if (selectProp) {
			const entry = editor.api.node({
				at: options.at,
				match: { type: editor.getType(KEYS.column) }
			});
			if (!entry) return;
			editor.tf.select(entry[1].concat([0]));
		}
	});
};

//#endregion
//#region src/lib/transforms/moveMiddleColumn.ts
/**
* Move the middle column to the left if direction is 'left', or to the right if
* 'right'. If the middle node is empty, return false and remove it.
*/
const moveMiddleColumn = (editor, [node, path], options) => {
	if ((options?.direction || "left") === "left") {
		const DESCENDANT_PATH = [1];
		const middleChildNode = NodeApi.get(node, DESCENDANT_PATH);
		if (!middleChildNode) return false;
		const isEmpty = NodeApi.string(middleChildNode) === "";
		const middleChildPathRef = editor.api.pathRef(path.concat(DESCENDANT_PATH));
		if (isEmpty) {
			editor.tf.removeNodes({ at: middleChildPathRef.current });
			return false;
		}
		const firstNode = NodeApi.descendant(node, [0]);
		if (!firstNode) return false;
		const firstLast = path.concat([0, firstNode.children.length]);
		editor.tf.moveNodes({
			at: middleChildPathRef.current,
			to: firstLast
		});
		editor.tf.unwrapNodes({ at: middleChildPathRef.current });
		middleChildPathRef.unref();
	}
};

//#endregion
//#region src/lib/transforms/resizeColumn.ts
function resizeColumn(columnGroup, columnId, newWidthPercent) {
	const widths = columnGroup.children.map((col) => col.width ? Number.parseFloat(col.width) : 0);
	if (widths.reduce((sum, w) => sum + w, 0) === 0) {
		const evenWidth = 100 / columnGroup.children.length;
		columnGroup.children.forEach((col) => {
			col.width = `${evenWidth}%`;
		});
		return columnGroup;
	}
	const index = columnGroup.children.findIndex((col) => col.id === columnId);
	if (index === -1) return columnGroup;
	widths[index] = newWidthPercent;
	let totalAfter = widths.reduce((sum, w) => sum + w, 0);
	const diff = 100 - totalAfter;
	if (diff !== 0) {
		const siblingIndex = (index + 1) % widths.length;
		widths[siblingIndex] = Math.max(widths[siblingIndex] + diff, 0);
	}
	totalAfter = widths.reduce((sum, w) => sum + w, 0);
	if (Math.round(totalAfter) !== 100) {
		const scale = 100 / totalAfter;
		for (let i = 0; i < widths.length; i++) widths[i] = Number.parseFloat((widths[i] * scale).toFixed(2));
	}
	columnGroup.children.forEach((col, i) => {
		col.width = `${widths[i]}%`;
	});
	return columnGroup;
}

//#endregion
//#region src/lib/utils/columnsToWidths.ts
const columnsToWidths = ({ columns = 2 } = {}) => new Array(columns).fill(null).map(() => `${100 / columns}%`);

//#endregion
//#region src/lib/transforms/setColumns.ts
const setColumns = (editor, { at, columns, widths }) => {
	editor.tf.withoutNormalizing(() => {
		if (!at) return;
		widths = widths ?? columnsToWidths({ columns });
		if (widths.length === 0) return;
		const columnGroup = editor.api.node(at);
		if (!columnGroup) return;
		const [{ children }, path] = columnGroup;
		const currentCount = children.length;
		const targetCount = widths.length;
		if (currentCount === targetCount) {
			widths.forEach((width, i) => {
				editor.tf.setNodes({ width }, { at: path.concat([i]) });
			});
			return;
		}
		if (targetCount > currentCount) {
			const columnsToAdd = targetCount - currentCount;
			const insertPath = path.concat([currentCount]);
			const newColumns = new Array(columnsToAdd).fill(null).map((_, i) => ({
				children: [editor.api.create.block()],
				type: editor.getType(KEYS.column),
				width: widths[currentCount + i] || `${100 / targetCount}%`
			}));
			editor.tf.insertNodes(newColumns, { at: insertPath });
			widths.forEach((width, i) => {
				editor.tf.setNodes({ width }, { at: path.concat([i]) });
			});
			return;
		}
		if (targetCount < currentCount) {
			const keepColumnIndex = targetCount - 1;
			const keepColumnPath = path.concat([keepColumnIndex]);
			const keepColumnNode = NodeApi.get(editor, keepColumnPath);
			if (!keepColumnNode) return;
			const to = keepColumnPath.concat([keepColumnNode.children.length]);
			for (let i = currentCount - 1; i > keepColumnIndex; i--) {
				const columnPath = path.concat([i]);
				const columnEntry = editor.api.node(columnPath);
				if (!columnEntry) continue;
				editor.tf.moveNodes({
					at: columnEntry[1],
					children: true,
					to
				});
			}
			for (let i = currentCount - 1; i > keepColumnIndex; i--) editor.tf.removeNodes({ at: path.concat([i]) });
			widths.forEach((width, i) => {
				editor.tf.setNodes({ width }, { at: path.concat([i]) });
			});
		}
	});
};

//#endregion
//#region src/lib/transforms/toggleColumnGroup.ts
const toggleColumnGroup = (editor, { at, columns = 2, widths } = {}) => {
	const entry = editor.api.block({ at });
	const columnGroupEntry = editor.api.block({
		above: true,
		at,
		match: { type: editor.getType(KEYS.columnGroup) }
	});
	if (!entry) return;
	const [node, path] = entry;
	if (columnGroupEntry) setColumns(editor, {
		at: columnGroupEntry[1],
		columns,
		widths
	});
	else {
		const columnWidths = widths || columnsToWidths({ columns });
		const nodes = {
			children: new Array(columns).fill(null).map((_, index) => ({
				children: [index === 0 ? node : editor.api.create.block()],
				type: editor.getType(KEYS.column),
				width: columnWidths[index]
			})),
			type: editor.getType(KEYS.columnGroup)
		};
		editor.tf.replaceNodes(nodes, { at: path });
		editor.tf.select(editor.api.start(path.concat([0])));
	}
};

//#endregion
export { BaseColumnItemPlugin, BaseColumnPlugin, columnsToWidths, insertColumn, insertColumnGroup, moveMiddleColumn, resizeColumn, setColumns, toggleColumnGroup, withColumn };
