// @platejs/math@53.0.0/dist/react/index.js
import { n as BaseEquationPlugin, t as BaseInlineEquationPlugin } from "../BaseInlineEquationPlugin-1MXNHk1G.js";
import { isHotkey } from "platejs";
import katex from "katex";
import { toPlatePlugin, useEditorRef, useElement } from "platejs/react";
import React, { useEffect, useRef } from "react";

//#region src/react/EquationPlugin.tsx
const EquationPlugin = toPlatePlugin(BaseEquationPlugin);

//#endregion
//#region src/react/InlineEquationPlugin.tsx
const InlineEquationPlugin = toPlatePlugin(BaseInlineEquationPlugin);

//#endregion
//#region src/react/hooks/useEquationElement.ts
const useEquationElement = ({ element, katexRef, options }) => {
	React.useEffect(() => {
		if (!katexRef.current) return;
		katex.render(element.texExpression, katexRef.current, options);
	}, [element.texExpression]);
};

//#endregion
//#region src/react/hooks/useEquationInput.ts
const useEquationInput = ({ isInline, open, onClose }) => {
	const editor = useEditorRef();
	const element = useElement();
	const inputRef = useRef(null);
	const [expressionInput, setExpressionInput] = React.useState(element.texExpression);
	const initialExpressionRef = useRef(element.texExpression);
	useEffect(() => {
		if (open) setTimeout(() => {
			if (inputRef.current) {
				inputRef.current.focus();
				inputRef.current.select();
				if (isInline) initialExpressionRef.current = element.texExpression;
			}
		}, 0);
	}, [open]);
	useEffect(() => {
		const setExpression = () => {
			editor.tf.setNodes({ texExpression: expressionInput || "" }, { at: element });
		};
		if (isInline) editor.tf.withMerging(setExpression);
		else setExpression();
	}, [expressionInput]);
	const onSubmit = () => {
		onClose?.();
	};
	const onDismiss = () => {
		if (isInline) editor.tf.setNodes({ texExpression: initialExpressionRef.current }, { at: element });
		onClose?.();
	};
	return {
		props: {
			value: expressionInput,
			onChange: (e) => {
				setExpressionInput(e.target.value);
			},
			onKeyDown: (e) => {
				if (isHotkey("enter")(e)) {
					e.preventDefault();
					onSubmit();
				} else if (isHotkey("escape")(e)) {
					e.preventDefault();
					onDismiss();
				}
				if (isInline) {
					const { selectionEnd, selectionStart, value } = e.target;
					if (selectionStart === 0 && selectionEnd === 0 && isHotkey("ArrowLeft")(e)) {
						e.preventDefault();
						editor.tf.select(element, {
							focus: true,
							previous: true
						});
					}
					if (selectionEnd === value.length && selectionStart === value.length && isHotkey("ArrowRight")(e)) {
						e.preventDefault();
						editor.tf.select(element, {
							focus: true,
							next: true
						});
					}
				}
			}
		},
		ref: inputRef,
		onDismiss,
		onSubmit
	};
};

//#endregion
export { EquationPlugin, InlineEquationPlugin, useEquationElement, useEquationInput };
