// @platejs/resizable@53.0.0/dist/index.js
import { c } from "react-compiler-runtime";
import React from "react";
import { createAtomStore, createPrimitiveComponent, useEditorRef, useElement, usePath, useReadOnly } from "platejs/react";

//#region src/utils/isTouchEvent.ts
const isTouchEvent = (event) => "touches" in event;

//#endregion
//#region src/utils/resizeLengthToRelative.ts
const resizeLengthToRelative = (length, parentLength) => {
	if (typeof length === "number") return `${length / parentLength * 100}%`;
	return length;
};

//#endregion
//#region src/utils/resizeLengthToStatic.ts
const resizeLengthToStatic = (length, parentLength) => {
	if (typeof length === "string") return parentLength * Number.parseFloat(length) / 100;
	return length;
};

//#endregion
//#region src/utils/resizeLengthClamp.ts
const resizeLengthClampStatic = (length, { max, min }) => {
	let result = length;
	if (min !== void 0) result = Math.max(result, min);
	if (max !== void 0) result = Math.min(result, max);
	return result;
};
const resizeLengthClamp = (length, parentLength, { max, min }) => {
	const clampedStaticLength = resizeLengthClampStatic(resizeLengthToStatic(length, parentLength), {
		max: max === void 0 ? void 0 : resizeLengthToStatic(max, parentLength),
		min: min === void 0 ? void 0 : resizeLengthToStatic(min, parentLength)
	});
	switch (typeof length) {
		case "number": return clampedStaticLength;
		case "string": return resizeLengthToRelative(clampedStaticLength, parentLength);
		default: throw new Error("Invalid length type");
	}
};

//#endregion
//#region src/components/ResizeHandle.tsx
const initialState = { onResize: null };
const { ResizeHandleProvider, useResizeHandleSet, useResizeHandleStore, useResizeHandleValue } = createAtomStore(initialState, {
	name: "resizeHandle",
	suppressWarnings: true
});
const useResizeHandleState = (t0) => {
	const $ = c(21);
	const { direction: t1, initialSize: _initialSize, onHover, onHoverEnd, onMouseDown, onResize: onResizeProp, onTouchStart } = t0;
	const direction = t1 === void 0 ? "left" : t1;
	const readOnly = useReadOnly();
	const onResizeStore = useResizeHandleValue("onResize");
	const onResize = onResizeProp ?? onResizeStore;
	const [isResizing, setIsResizing] = React.useState(false);
	const [initialPosition, setInitialPosition] = React.useState(0);
	const [initialSizeState, setInitialSize] = React.useState(0);
	const initialSize = _initialSize ?? initialSizeState;
	const isHorizontal = direction === "left" || direction === "right";
	let t2;
	let t3;
	if ($[0] !== direction || $[1] !== initialPosition || $[2] !== initialSize || $[3] !== isHorizontal || $[4] !== isResizing || $[5] !== onHoverEnd || $[6] !== onResize) {
		t2 = () => {
			if (!isResizing) return;
			const sendResizeEvent = (event, finished) => {
				const { clientX, clientY } = isTouchEvent(event) ? event.touches[0] || event.changedTouches[0] : event;
				onResize({
					delta: (isHorizontal ? clientX : clientY) - initialPosition,
					direction,
					finished,
					initialSize
				});
			};
			const handleMouseMove = (event_0) => sendResizeEvent(event_0, false);
			const handleMouseUp = (event_1) => {
				setIsResizing(false);
				onHoverEnd?.();
				sendResizeEvent(event_1, true);
			};
			window.addEventListener("mousemove", handleMouseMove);
			window.addEventListener("mouseup", handleMouseUp);
			window.addEventListener("touchmove", handleMouseMove);
			window.addEventListener("touchend", handleMouseUp);
			return () => {
				window.removeEventListener("mousemove", handleMouseMove);
				window.removeEventListener("mouseup", handleMouseUp);
				window.removeEventListener("touchmove", handleMouseMove);
				window.removeEventListener("touchend", handleMouseUp);
			};
		};
		t3 = [
			isResizing,
			initialPosition,
			initialSize,
			onResize,
			isHorizontal,
			onHoverEnd,
			direction
		];
		$[0] = direction;
		$[1] = initialPosition;
		$[2] = initialSize;
		$[3] = isHorizontal;
		$[4] = isResizing;
		$[5] = onHoverEnd;
		$[6] = onResize;
		$[7] = t2;
		$[8] = t3;
	} else {
		t2 = $[7];
		t3 = $[8];
	}
	React.useEffect(t2, t3);
	let t4;
	if ($[9] !== direction || $[10] !== initialPosition || $[11] !== initialSize || $[12] !== isHorizontal || $[13] !== isResizing || $[14] !== onHover || $[15] !== onHoverEnd || $[16] !== onMouseDown || $[17] !== onResize || $[18] !== onTouchStart || $[19] !== readOnly) {
		t4 = {
			direction,
			initialPosition,
			initialSize,
			isHorizontal,
			isResizing,
			readOnly,
			setInitialPosition,
			setInitialSize,
			setIsResizing,
			onHover,
			onHoverEnd,
			onMouseDown,
			onResize,
			onTouchStart
		};
		$[9] = direction;
		$[10] = initialPosition;
		$[11] = initialSize;
		$[12] = isHorizontal;
		$[13] = isResizing;
		$[14] = onHover;
		$[15] = onHoverEnd;
		$[16] = onMouseDown;
		$[17] = onResize;
		$[18] = onTouchStart;
		$[19] = readOnly;
		$[20] = t4;
	} else t4 = $[20];
	return t4;
};
const useResizeHandle = ({ isHorizontal, isResizing, readOnly, setInitialPosition, setInitialSize, setIsResizing, onHover, onHoverEnd, onMouseDown, onTouchStart }) => {
	const handleMouseDown = (event) => {
		const { clientX, clientY } = event;
		setInitialPosition(isHorizontal ? clientX : clientY);
		const element = event.target.parentElement;
		setInitialSize(isHorizontal ? element.offsetWidth : element.offsetHeight);
		setIsResizing(true);
		onMouseDown?.(event);
	};
	const handleTouchStart = (event) => {
		const { touches } = event;
		const { clientX, clientY } = touches[0];
		setInitialPosition(isHorizontal ? clientX : clientY);
		const element = event.target.parentElement;
		setInitialSize(isHorizontal ? element.offsetWidth : element.offsetHeight);
		setIsResizing(true);
		onTouchStart?.(event);
	};
	const handleMouseOver = () => {
		onHover?.();
	};
	const handleMouseOut = () => {
		if (!isResizing) onHoverEnd?.();
	};
	return {
		hidden: readOnly,
		props: {
			onMouseDown: handleMouseDown,
			onMouseOut: handleMouseOut,
			onMouseOver: handleMouseOver,
			onTouchEnd: handleMouseOut,
			onTouchMove: handleMouseOver,
			onTouchStart: handleTouchStart
		}
	};
};
const ResizeHandle = createPrimitiveComponent("div")({
	propsHook: useResizeHandle,
	stateHook: useResizeHandleState
});

//#endregion
//#region src/components/useResizableStore.ts
const { ResizableProvider, resizableStore, useResizableSet, useResizableStore, useResizableValue } = createAtomStore({ width: 0 }, { name: "resizable" });

//#endregion
//#region src/components/Resizable.tsx
const useResizableState = (t0) => {
	const $ = c(17);
	let t1;
	if ($[0] !== t0) {
		t1 = t0 === void 0 ? {} : t0;
		$[0] = t0;
		$[1] = t1;
	} else t1 = $[1];
	const { align: t2, maxWidth: t3, minWidth: t4 } = t1;
	const align = t2 === void 0 ? "center" : t2;
	const maxWidth = t3 === void 0 ? "100%" : t3;
	const minWidth = t4 === void 0 ? 92 : t4;
	const editor = useEditorRef();
	const element = useElement();
	const path = usePath();
	const nodeWidth = element?.width ?? "100%";
	const width = useResizableValue("width");
	const setWidth = useResizableSet("width");
	let t5;
	if ($[2] !== editor || $[3] !== nodeWidth || $[4] !== path) {
		t5 = (w) => {
			if (w === nodeWidth) editor.tf.select(path);
			else editor.tf.setNodes({ width: w }, { at: path });
		};
		$[2] = editor;
		$[3] = nodeWidth;
		$[4] = path;
		$[5] = t5;
	} else t5 = $[5];
	const setNodeWidth = t5;
	let t6;
	let t7;
	if ($[6] !== nodeWidth || $[7] !== setWidth) {
		t6 = () => {
			setWidth(nodeWidth);
		};
		t7 = [nodeWidth, setWidth];
		$[6] = nodeWidth;
		$[7] = setWidth;
		$[8] = t6;
		$[9] = t7;
	} else {
		t6 = $[8];
		t7 = $[9];
	}
	React.useEffect(t6, t7);
	let t8;
	if ($[10] !== align || $[11] !== maxWidth || $[12] !== minWidth || $[13] !== setNodeWidth || $[14] !== setWidth || $[15] !== width) {
		t8 = {
			align,
			maxWidth,
			minWidth,
			setNodeWidth,
			setWidth,
			width
		};
		$[10] = align;
		$[11] = maxWidth;
		$[12] = minWidth;
		$[13] = setNodeWidth;
		$[14] = setWidth;
		$[15] = width;
		$[16] = t8;
	} else t8 = $[16];
	return t8;
};
const useResizable = (t0) => {
	const $ = c(19);
	const { align, maxWidth, minWidth, setNodeWidth, setWidth, width } = t0;
	const wrapperRef = React.useRef(null);
	let t1;
	if ($[0] !== align || $[1] !== maxWidth || $[2] !== minWidth || $[3] !== setNodeWidth || $[4] !== setWidth) {
		t1 = (t2$1) => {
			const { delta, direction, finished, initialSize } = t2$1;
			const wrapperStaticWidth = wrapperRef.current.offsetWidth;
			const newWidth = resizeLengthClamp(initialSize + delta * ((align === "center" ? 2 : 1) * (direction === "left" ? -1 : 1)), wrapperStaticWidth, {
				max: maxWidth,
				min: minWidth
			});
			if (finished) setNodeWidth(newWidth);
			else setWidth(newWidth);
		};
		$[0] = align;
		$[1] = maxWidth;
		$[2] = minWidth;
		$[3] = setNodeWidth;
		$[4] = setWidth;
		$[5] = t1;
	} else t1 = $[5];
	const t2 = t1;
	let t3;
	if ($[6] !== t2) {
		t3 = { onResize: t2 };
		$[6] = t2;
		$[7] = t3;
	} else t3 = $[7];
	let t4;
	if ($[8] !== maxWidth || $[9] !== minWidth || $[10] !== width) {
		t4 = {
			maxWidth,
			minWidth,
			position: "relative",
			width
		};
		$[8] = maxWidth;
		$[9] = minWidth;
		$[10] = width;
		$[11] = t4;
	} else t4 = $[11];
	const t5 = t4;
	let t6;
	if ($[12] !== t5) {
		t6 = { style: t5 };
		$[12] = t5;
		$[13] = t6;
	} else t6 = $[13];
	let t7;
	if ($[14] === Symbol.for("react.memo_cache_sentinel")) {
		t7 = { position: "relative" };
		$[14] = t7;
	} else t7 = $[14];
	let t8;
	if ($[15] === Symbol.for("react.memo_cache_sentinel")) {
		t8 = { style: t7 };
		$[15] = t8;
	} else t8 = $[15];
	let t9;
	if ($[16] !== t3 || $[17] !== t6) {
		t9 = {
			context: t3,
			props: t6,
			wrapperProps: t8,
			wrapperRef
		};
		$[16] = t3;
		$[17] = t6;
		$[18] = t9;
	} else t9 = $[18];
	return t9;
};
const Resizable = React.forwardRef((t0, ref) => {
	const $ = c(16);
	let children;
	let options;
	let rest;
	if ($[0] !== t0) {
		({children, options, ...rest} = t0);
		$[0] = t0;
		$[1] = children;
		$[2] = options;
		$[3] = rest;
	} else {
		children = $[1];
		options = $[2];
		rest = $[3];
	}
	const { context, props, wrapperProps, wrapperRef } = useResizable(useResizableState(options));
	let t1;
	if ($[4] !== children || $[5] !== context.onResize) {
		t1 = /* @__PURE__ */ React.createElement(ResizeHandleProvider, { onResize: context.onResize }, children);
		$[4] = children;
		$[5] = context.onResize;
		$[6] = t1;
	} else t1 = $[6];
	let t2;
	if ($[7] !== props || $[8] !== ref || $[9] !== rest || $[10] !== t1) {
		t2 = /* @__PURE__ */ React.createElement("div", {
			ref,
			...props,
			...rest
		}, t1);
		$[7] = props;
		$[8] = ref;
		$[9] = rest;
		$[10] = t1;
		$[11] = t2;
	} else t2 = $[11];
	let t3;
	if ($[12] !== t2 || $[13] !== wrapperProps || $[14] !== wrapperRef) {
		t3 = /* @__PURE__ */ React.createElement("div", {
			ref: wrapperRef,
			...wrapperProps
		}, t2);
		$[12] = t2;
		$[13] = wrapperProps;
		$[14] = wrapperRef;
		$[15] = t3;
	} else t3 = $[15];
	return t3;
});
Resizable.displayName = "Resizable";

//#endregion
export { Resizable, ResizableProvider, ResizeHandle, ResizeHandleProvider, isTouchEvent, resizableStore, resizeLengthClamp, resizeLengthClampStatic, resizeLengthToRelative, resizeLengthToStatic, useResizable, useResizableSet, useResizableState, useResizableStore, useResizableValue, useResizeHandle, useResizeHandleSet, useResizeHandleState, useResizeHandleStore, useResizeHandleValue };
