// @platejs/media@53.1.4/dist/src-C28rUpXn.js
import { KEYS, bindFirst, createSlatePlugin, createTSlatePlugin, getInjectedPlugins, isUrl, pipeInsertDataQuery } from "platejs";

//#region src/lib/BaseAudioPlugin.ts
const BaseAudioPlugin = createSlatePlugin({
	key: KEYS.audio,
	node: {
		isElement: true,
		isVoid: true
	}
});

//#endregion
//#region src/lib/BaseFilePlugin.ts
const BaseFilePlugin = createSlatePlugin({
	key: KEYS.file,
	node: {
		isElement: true,
		isVoid: true
	}
});

//#endregion
//#region src/lib/BaseVideoPlugin.ts
const BaseVideoPlugin = createSlatePlugin({
	key: KEYS.video,
	node: {
		dangerouslyAllowAttributes: ["width", "height"],
		isElement: true,
		isVoid: true
	}
});

//#endregion
//#region src/lib/image/transforms/insertImage.ts
const insertImage = (editor, url, options = {}) => {
	const image = {
		children: [{ text: "" }],
		type: editor.getType(KEYS.img),
		url
	};
	editor.tf.insertNodes(image, {
		nextBlock: true,
		...options
	});
};

//#endregion
//#region src/lib/image/transforms/insertImageFromFiles.ts
const insertImageFromFiles = (editor, files, options = {}) => {
	for (const file of files) {
		const reader = new FileReader();
		const [mime] = file.type.split("/");
		if (mime === "image") {
			reader.addEventListener("load", async () => {
				if (!reader.result) return;
				const uploadImage = editor.getOptions(BaseImagePlugin).uploadImage;
				insertImage(editor, uploadImage ? await uploadImage(reader.result) : reader.result, options);
			});
			reader.readAsDataURL(file);
		}
	}
};

//#endregion
//#region src/lib/image/utils/isImageUrl.ts
const imageExtensions = new Set([
	"3dv",
	"ai",
	"amf",
	"art",
	"art",
	"ase",
	"awg",
	"blp",
	"bmp",
	"bw",
	"bw",
	"cd5",
	"cdr",
	"cgm",
	"cit",
	"cmx",
	"cpt",
	"cr2",
	"cur",
	"cut",
	"dds",
	"dib",
	"djvu",
	"dxf",
	"e2d",
	"ecw",
	"egt",
	"egt",
	"emf",
	"eps",
	"exif",
	"fs",
	"gbr",
	"gif",
	"gpl",
	"grf",
	"hdp",
	"icns",
	"ico",
	"iff",
	"iff",
	"int",
	"int",
	"inta",
	"jfif",
	"jng",
	"jp2",
	"jpeg",
	"jpg",
	"jps",
	"jxr",
	"lbm",
	"lbm",
	"liff",
	"max",
	"miff",
	"mng",
	"msp",
	"nitf",
	"nrrd",
	"odg",
	"ota",
	"pam",
	"pbm",
	"pc1",
	"pc2",
	"pc3",
	"pcf",
	"pct",
	"pcx",
	"pcx",
	"pdd",
	"pdn",
	"pgf",
	"pgm",
	"PI1",
	"PI2",
	"PI3",
	"pict",
	"png",
	"pnm",
	"pns",
	"ppm",
	"psb",
	"psd",
	"psp",
	"px",
	"pxm",
	"pxr",
	"qfx",
	"ras",
	"raw",
	"rgb",
	"rgb",
	"rgba",
	"rle",
	"sct",
	"sgi",
	"sgi",
	"sid",
	"stl",
	"sun",
	"svg",
	"sxd",
	"tga",
	"tga",
	"tif",
	"tiff",
	"v2d",
	"vnd",
	"vrml",
	"vtf",
	"wdp",
	"webp",
	"wmf",
	"x3d",
	"xar",
	"xbm",
	"xcf",
	"xpm"
]);
const isImageUrl = (url) => {
	if (!isUrl(url)) return false;
	const ext = new URL(url).pathname.split(".").pop();
	return imageExtensions.has(ext);
};

//#endregion
//#region src/lib/image/withImageEmbed.ts
/** If inserted text is image url, insert image instead. */
const withImageEmbed = ({ editor, getOptions, tf: { insertData } }) => ({ transforms: { insertData(dataTransfer) {
	if (getOptions().disableEmbedInsert) return insertData(dataTransfer);
	const text = dataTransfer.getData("text/plain");
	if (isImageUrl(text)) {
		insertImage(editor, text);
		return;
	}
	insertData(dataTransfer);
} } });

//#endregion
//#region src/lib/image/withImageUpload.ts
/**
* Allows for pasting images from clipboard. Not yet: dragging and dropping
* images, selecting them through a file system dialog.
*/
const withImageUpload = ({ editor, getOptions, plugin, tf: { insertData } }) => ({ transforms: { insertData(dataTransfer) {
	if (getOptions().disableUploadInsert) return insertData(dataTransfer);
	const mimeType = "text/plain";
	const text = dataTransfer.getData(mimeType);
	const { files } = dataTransfer;
	if (!text && files && files.length > 0) {
		if (!pipeInsertDataQuery(editor, getInjectedPlugins(editor, plugin), {
			data: text,
			dataTransfer,
			mimeType
		})) return insertData(dataTransfer);
		insertImageFromFiles(editor, files);
	} else return insertData(dataTransfer);
} } });

//#endregion
//#region src/lib/image/BaseImagePlugin.ts
/** Enables support for images. */
const BaseImagePlugin = createTSlatePlugin({
	key: KEYS.img,
	node: {
		dangerouslyAllowAttributes: [
			"alt",
			"width",
			"height"
		],
		isElement: true,
		isVoid: true
	},
	parsers: { html: { deserializer: {
		rules: [{ validNodeName: "IMG" }],
		parse: ({ element, type }) => ({
			type,
			url: element.getAttribute("src")
		})
	} } }
}).overrideEditor(withImageUpload).overrideEditor(withImageEmbed).extendEditorTransforms(({ editor }) => ({ insert: { imageFromFiles: bindFirst(insertImageFromFiles, editor) } }));

//#endregion
//#region src/lib/media/insertMedia.ts
const insertMedia = async (editor, { getUrl, type = editor.getType(KEYS.img), ...options } = {}) => {
	const url = getUrl ? await getUrl() : window.prompt(`Enter the URL of the ${type === KEYS.img ? KEYS.img : KEYS.mediaEmbed}`);
	if (!url) return;
	if (type === editor.getType(KEYS.img)) insertImage(editor, url, options);
	else insertMediaEmbed(editor, { url }, options);
};

//#endregion
//#region src/lib/media/parseMediaUrl.ts
const allowedProtocols = new Set(["http:", "https:"]);
const parseMediaUrl = (url, { urlParsers }) => {
	const embed = (() => {
		for (const parser of urlParsers) {
			const data = parser(url);
			if (data) return data;
		}
	})();
	if (embed?.url) try {
		const { protocol } = new URL(embed.url);
		if (!allowedProtocols.has(protocol)) return;
	} catch {
		console.warn(`Could not parse URL: ${embed.url}`);
		return;
	}
	return embed;
};

//#endregion
//#region src/lib/media-embed/parseIframeUrl.ts
const regexMatchIframeSrc = /<iframe\b[^>]*\bsrc="([^"]*)"/i;
const regexGroupQuotes = /"([^"]*)"/;
const regexTwitterStatusUrl = /https?:\/\/(?:twitter|x)\.com\/(?:#!\/)?\w+\/status(?:es)?\/\d+/;
const parseIframeUrl = (url) => {
	if (!url.startsWith("http")) {
		const twitterStatusUrl = url.match(regexTwitterStatusUrl)?.[0];
		if (twitterStatusUrl) return twitterStatusUrl;
		const returnString = (regexMatchIframeSrc.exec(url)?.[0])?.match(regexGroupQuotes)?.[1];
		if (returnString) return returnString;
	}
	return url;
};

//#endregion
//#region src/lib/media-embed/BaseMediaEmbedPlugin.ts
/**
* Enables support for embeddable media such as YouTube or Vimeo videos,
* Instagram posts and tweets or Google Maps.
*/
const BaseMediaEmbedPlugin = createTSlatePlugin({
	key: KEYS.mediaEmbed,
	node: {
		isElement: true,
		isVoid: true
	},
	options: { transformUrl: parseIframeUrl },
	parsers: { html: { deserializer: {
		rules: [{ validNodeName: "IFRAME" }],
		parse: ({ element, type }) => {
			const url = element.getAttribute("src");
			if (url) return {
				type,
				url
			};
		}
	} } }
});

//#endregion
//#region src/lib/media-embed/parseTwitterUrl.ts
const twitterRegex = /^https?:\/\/(?:twitter|x)\.com\/(?:#!\/)?(\w+)\/status(es)?\/(\d+)/;
const TWITTER_ID_INDEX = 3;
const parseTwitterUrl = (url) => {
	const match = twitterRegex.exec(url);
	if (match) return {
		id: match[TWITTER_ID_INDEX],
		provider: "twitter",
		sourceKind: "url",
		url
	};
};

//#endregion
//#region src/lib/media-embed/parseVideoUrl.ts
const YOUTUBE_PREFIX = "https://www.youtube.com/embed/";
const VIMEO_PREFIX = "https://player.vimeo.com/video/";
const DAILYMOTION_PREFIX = "https://www.dailymotion.com/embed/video/";
const COUB_PREFIX = "https://coub.com/embed/";
const YOUKU_PREFIX = "https://player.youku.com/embed/";
const VIDEO_PROVIDER_VALUES = [
	"youtube",
	"vimeo",
	"dailymotion",
	"youku",
	"coub"
];
const VIDEO_PROVIDERS = [...VIDEO_PROVIDER_VALUES];
const providerUrls = {
	coub: COUB_PREFIX,
	dailymotion: DAILYMOTION_PREFIX,
	vimeo: VIMEO_PREFIX,
	youku: YOUKU_PREFIX,
	youtube: YOUTUBE_PREFIX
};
const parseUrl = (url) => {
	if (!isUrl(url)) return;
	try {
		return new URL(url);
	} catch {
		return;
	}
};
const getPathSegments = (url) => url.pathname.split("/").filter(Boolean);
const getSegmentAfter = (segments, value) => {
	const index = segments.indexOf(value);
	return index >= 0 ? segments[index + 1] : void 0;
};
const isAlphaNumeric = (value) => value.length > 0 && [...value].every((char) => {
	const code = char.charCodeAt(0);
	return code >= 48 && code <= 57 || code >= 65 && code <= 90 || code >= 97 && code <= 122;
});
const isNumeric = (value) => value.length > 0 && [...value].every((char) => {
	const code = char.charCodeAt(0);
	return code >= 48 && code <= 57;
});
const isYoutubeId = (value) => value?.length === 11 && [...value].every((char) => {
	const code = char.charCodeAt(0);
	return char === "-" || char === "_" || code >= 48 && code <= 57 || code >= 65 && code <= 90 || code >= 97 && code <= 122;
});
const stripNonAlphaNumericSuffix = (value) => {
	if (!value) return;
	const id = [...value].findIndex((char) => !isAlphaNumeric(char));
	return id === -1 ? value : value.slice(0, id);
};
const parseYoutubeId = (url) => {
	const host = url.hostname.toLowerCase();
	const segments = getPathSegments(url);
	if (host === "youtu.be") {
		const id = segments[0];
		return isYoutubeId(id) ? id : void 0;
	}
	if (host !== "youtube.com" && !host.endsWith(".youtube.com")) return;
	const queryId = url.searchParams.get("v") ?? url.searchParams.get("ci");
	if (isYoutubeId(queryId ?? void 0)) return queryId;
	const pathId = getSegmentAfter(segments, "embed") ?? getSegmentAfter(segments, "v") ?? getSegmentAfter(segments, "vi") ?? getSegmentAfter(segments, "videos") ?? getSegmentAfter(segments, "shorts") ?? getSegmentAfter(segments, "live");
	return isYoutubeId(pathId) ? pathId : void 0;
};
const parseVimeoId = (url) => {
	const host = url.hostname.toLowerCase();
	if (host !== "vimeo.com" && !host.endsWith(".vimeo.com") && host !== "vimeopro.com" && !host.endsWith(".vimeopro.com")) return;
	return [...getPathSegments(url)].reverse().find(isNumeric);
};
const parseDailymotionId = (url) => {
	const host = url.hostname.toLowerCase();
	if (host !== "dai.ly" && host !== "dailymotion.com" && !host.endsWith(".dailymotion.com")) return;
	const segments = getPathSegments(url);
	const id = stripNonAlphaNumericSuffix(host === "dai.ly" ? segments[0] : getSegmentAfter(segments, "video"));
	return id && isAlphaNumeric(id) ? id : void 0;
};
const parseYoukuId = (url) => {
	const host = url.hostname.toLowerCase();
	if (host !== "youku.com" && !host.endsWith(".youku.com")) return;
	const segments = getPathSegments(url);
	const videoIdParam = url.searchParams.get("VideoIDS") ?? void 0;
	const pathId = getSegmentAfter(segments, "embed") ?? getSegmentAfter(segments, "sid") ?? segments.find((segment) => segment.startsWith("id_"))?.slice(3).split(".")[0];
	const id = stripNonAlphaNumericSuffix(videoIdParam ?? pathId);
	return id && isAlphaNumeric(id) ? id : void 0;
};
const parseCoubId = (url) => {
	const host = url.hostname.toLowerCase();
	if (host !== "coub.com" && !host.endsWith(".coub.com")) return;
	const segments = getPathSegments(url);
	const id = getSegmentAfter(segments, "view") ?? getSegmentAfter(segments, "embed");
	return id && isAlphaNumeric(id) ? id : void 0;
};
const parsers = {
	coub: parseCoubId,
	dailymotion: parseDailymotionId,
	vimeo: parseVimeoId,
	youku: parseYoukuId,
	youtube: parseYoutubeId
};
const parseVideoUrl = (url) => {
	const parsedUrl = parseUrl(url);
	if (!parsedUrl) return;
	for (const provider of VIDEO_PROVIDER_VALUES) {
		const id = parsers[provider](parsedUrl);
		if (!id) continue;
		const embedUrl = `${providerUrls[provider]}${id}`;
		return {
			id,
			provider,
			sourceKind: "url",
			sourceUrl: embedUrl === url ? void 0 : url,
			url: embedUrl
		};
	}
};

//#endregion
//#region src/lib/media-embed/transforms/insertMediaEmbed.ts
const insertMediaEmbed = (editor, { url = "" }, options = {}) => {
	if (!editor.selection) return;
	const selectionParentEntry = editor.api.parent(editor.selection);
	if (!selectionParentEntry) return;
	const [, path] = selectionParentEntry;
	const normalized = parseMediaUrl(url, { urlParsers: [parseTwitterUrl, parseVideoUrl] });
	editor.tf.insertNodes({
		children: [{ text: "" }],
		id: normalized?.id,
		provider: normalized?.provider,
		sourceUrl: normalized?.sourceUrl,
		type: editor.getType(KEYS.mediaEmbed),
		url: normalized?.url ?? url
	}, {
		at: path,
		nextBlock: true,
		...options
	});
};

//#endregion
//#region src/lib/placeholder/transforms/insertPlaceholder.ts
const insertPlaceholder = (editor, mediaType, options) => {
	editor.tf.withoutNormalizing(() => editor.tf.insertNodes({
		children: [{ text: "" }],
		mediaType,
		type: editor.getType(KEYS.placeholder)
	}, options));
};
const insertImagePlaceholder = (editor, options) => insertPlaceholder(editor, KEYS.img, options);
const insertVideoPlaceholder = (editor, options) => insertPlaceholder(editor, KEYS.video, options);
const insertAudioPlaceholder = (editor, options) => insertPlaceholder(editor, KEYS.audio, options);
const insertFilePlaceholder = (editor, options) => insertPlaceholder(editor, KEYS.file, options);

//#endregion
//#region src/lib/placeholder/transforms/setMediaNode.ts
const setMediaNode = (editor, props, options) => editor.tf.setNodes(props, options);

//#endregion
//#region src/lib/placeholder/BasePlaceholderPlugin.ts
const BasePlaceholderPlugin = createTSlatePlugin({
	key: KEYS.placeholder,
	node: {
		isElement: true,
		isVoid: true
	}
}).extendEditorTransforms(({ editor }) => ({ insert: {
	audioPlaceholder: bindFirst(insertAudioPlaceholder, editor),
	filePlaceholder: bindFirst(insertFilePlaceholder, editor),
	imagePlaceholder: bindFirst(insertImagePlaceholder, editor),
	videoPlaceholder: bindFirst(insertVideoPlaceholder, editor)
} }));

//#endregion
export { BaseFilePlugin as C, BaseVideoPlugin as S, withImageUpload as _, insertImagePlaceholder as a, insertImageFromFiles as b, insertMediaEmbed as c, parseTwitterUrl as d, BaseMediaEmbedPlugin as f, BaseImagePlugin as g, insertMedia as h, insertFilePlaceholder as i, VIDEO_PROVIDERS as l, parseMediaUrl as m, setMediaNode as n, insertPlaceholder as o, parseIframeUrl as p, insertAudioPlaceholder as r, insertVideoPlaceholder as s, BasePlaceholderPlugin as t, parseVideoUrl as u, withImageEmbed as v, BaseAudioPlugin as w, insertImage as x, isImageUrl as y };
