// @platejs/code-block@53.0.0/dist/BaseCodeBlockPlugin-stnOioFF.js
import { ElementApi, KEYS, NodeApi, createSlatePlugin, createTSlatePlugin } from "platejs";

//#region src/lib/deserializer/htmlDeserializerCodeBlock.ts
const htmlDeserializerCodeBlock = {
	rules: [{ validNodeName: "PRE" }, {
		validNodeName: "P",
		validStyle: { fontFamily: "Consolas" }
	}],
	parse: ({ element }) => {
		const languageSelectorText = [...element.childNodes].find((node) => node.nodeName === "SELECT")?.textContent || "";
		const textContent = element.textContent?.replace(languageSelectorText, "") || "";
		let lines = textContent.split("\n");
		if (!lines?.length) lines = [textContent];
		return {
			children: lines.map((line) => ({
				children: [{ text: line }],
				type: KEYS.codeLine
			})),
			type: KEYS.codeBlock
		};
	}
};

//#endregion
//#region src/lib/queries/getCodeLineEntry.ts
/** If at (default = selection) is in ul>li>p, return li and ul node entries. */
const getCodeLineEntry = (editor, { at = editor.selection } = {}) => {
	if (at && editor.api.some({
		at,
		match: { type: editor.getType(KEYS.codeLine) }
	})) {
		const selectionParent = editor.api.parent(at);
		if (!selectionParent) return;
		const [, parentPath] = selectionParent;
		const codeLine = editor.api.above({
			at,
			match: { type: editor.getType(KEYS.codeLine) }
		}) || editor.api.parent(parentPath);
		if (!codeLine) return;
		const [codeLineNode, codeLinePath] = codeLine;
		if (ElementApi.isElement(codeLineNode) && codeLineNode.type !== editor.getType(KEYS.codeLine)) return;
		const codeBlock = editor.api.parent(codeLinePath);
		if (!codeBlock) return;
		return {
			codeBlock,
			codeLine
		};
	}
};

//#endregion
//#region src/lib/queries/getIndentDepth.ts
const nonWhitespaceOrEndRegex = /\S|$/;
const getIndentDepth = (editor, { codeLine }) => {
	const [, codeLinePath] = codeLine;
	return editor.api.string(codeLinePath).search(nonWhitespaceOrEndRegex);
};

//#endregion
//#region src/lib/queries/isCodeBlockEmpty.ts
/** Is the selection inside an empty code block */
const isCodeBlockEmpty = (editor) => {
	const { codeBlock } = getCodeLineEntry(editor) ?? {};
	if (!codeBlock) return false;
	const codeLines = Array.from(NodeApi.children(editor, codeBlock[1]));
	if (codeLines.length === 0) return true;
	if (codeLines.length > 1) return false;
	const firstCodeLineNode = codeLines[0][0];
	return !NodeApi.string(firstCodeLineNode);
};

//#endregion
//#region src/lib/ensureStablePythonGrammar.ts
const STABLE_PYTHON_ALIASES = [
	"py",
	"gyp",
	"ipython"
];
const patchedLowlights = /* @__PURE__ */ new WeakSet();
function source(re) {
	if (!re) return null;
	if (typeof re === "string") return re;
	return re.source;
}
function lookahead(re) {
	return concat("(?=", re, ")");
}
function concat(...args) {
	return args.map((value) => source(value)).join("");
}
function pythonBrowserSafe(hljs) {
	const keywords = {
		$pattern: /[A-Za-z]\w+|__\w+__/,
		built_in: [
			"__import__",
			"abs",
			"all",
			"any",
			"ascii",
			"bin",
			"bool",
			"breakpoint",
			"bytearray",
			"bytes",
			"callable",
			"chr",
			"classmethod",
			"compile",
			"complex",
			"delattr",
			"dict",
			"dir",
			"divmod",
			"enumerate",
			"eval",
			"exec",
			"filter",
			"float",
			"format",
			"frozenset",
			"getattr",
			"globals",
			"hasattr",
			"hash",
			"help",
			"hex",
			"id",
			"input",
			"int",
			"isinstance",
			"issubclass",
			"iter",
			"len",
			"list",
			"locals",
			"map",
			"max",
			"memoryview",
			"min",
			"next",
			"object",
			"oct",
			"open",
			"ord",
			"pow",
			"print",
			"property",
			"range",
			"repr",
			"reversed",
			"round",
			"set",
			"setattr",
			"slice",
			"sorted",
			"staticmethod",
			"str",
			"sum",
			"super",
			"tuple",
			"type",
			"vars",
			"zip"
		],
		keyword: [
			"and",
			"as",
			"assert",
			"async",
			"await",
			"break",
			"case",
			"class",
			"continue",
			"def",
			"del",
			"elif",
			"else",
			"except",
			"finally",
			"for",
			"from",
			"global",
			"if",
			"import",
			"in",
			"is",
			"lambda",
			"match",
			"nonlocal|10",
			"not",
			"or",
			"pass",
			"raise",
			"return",
			"try",
			"while",
			"with",
			"yield"
		],
		literal: [
			"__debug__",
			"Ellipsis",
			"False",
			"None",
			"NotImplemented",
			"True"
		],
		type: [
			"Any",
			"Callable",
			"Coroutine",
			"Dict",
			"List",
			"Literal",
			"Generic",
			"Optional",
			"Sequence",
			"Set",
			"Tuple",
			"Type",
			"Union"
		]
	};
	const prompt = {
		begin: /^(>>>|\.\.\.) /,
		className: "meta"
	};
	const subst = {
		begin: /\{/,
		className: "subst",
		end: /\}/,
		illegal: /#/,
		keywords
	};
	const literalBracket = {
		begin: /\{\{/,
		relevance: 0
	};
	const string = {
		className: "string",
		contains: [hljs.BACKSLASH_ESCAPE],
		variants: [
			{
				begin: /([uU]|[bB]|[rR]|[bB][rR]|[rR][bB])?'''/,
				contains: [hljs.BACKSLASH_ESCAPE, prompt],
				end: /'''/,
				relevance: 10
			},
			{
				begin: /([uU]|[bB]|[rR]|[bB][rR]|[rR][bB])?"""/,
				contains: [hljs.BACKSLASH_ESCAPE, prompt],
				end: /"""/,
				relevance: 10
			},
			{
				begin: /([fF][rR]|[rR][fF]|[fF])'''/,
				contains: [
					hljs.BACKSLASH_ESCAPE,
					prompt,
					literalBracket,
					subst
				],
				end: /'''/
			},
			{
				begin: /([fF][rR]|[rR][fF]|[fF])"""/,
				contains: [
					hljs.BACKSLASH_ESCAPE,
					prompt,
					literalBracket,
					subst
				],
				end: /"""/
			},
			{
				begin: /([uU]|[rR])'/,
				end: /'/,
				relevance: 10
			},
			{
				begin: /([uU]|[rR])"/,
				end: /"/,
				relevance: 10
			},
			{
				begin: /([bB]|[bB][rR]|[rR][bB])'/,
				end: /'/
			},
			{
				begin: /([bB]|[bB][rR]|[rR][bB])"/,
				end: /"/
			},
			{
				begin: /([fF][rR]|[rR][fF]|[fF])'/,
				contains: [
					hljs.BACKSLASH_ESCAPE,
					literalBracket,
					subst
				],
				end: /'/
			},
			{
				begin: /([fF][rR]|[rR][fF]|[fF])"/,
				contains: [
					hljs.BACKSLASH_ESCAPE,
					literalBracket,
					subst
				],
				end: /"/
			},
			hljs.APOS_STRING_MODE,
			hljs.QUOTE_STRING_MODE
		]
	};
	const digitPart = "[0-9](_?[0-9])*";
	const pointFloat = `(\\b(${digitPart}))?\\.(${digitPart})|\\b(${digitPart})\\.`;
	const number = {
		className: "number",
		relevance: 0,
		variants: [
			{ begin: `(\\b(${digitPart})|(${pointFloat}))[eE][+-]?(${digitPart})[jJ]?\\b` },
			{ begin: `(${pointFloat})[jJ]?` },
			{ begin: "\\b([1-9](_?[0-9])*|0+(_?0)*)[lLjJ]?\\b" },
			{ begin: "\\b0[bB](_?[01])+[lL]?\\b" },
			{ begin: "\\b0[oO](_?[0-7])+[lL]?\\b" },
			{ begin: "\\b0[xX](_?[0-9a-fA-F])+[lL]?\\b" },
			{ begin: `\\b(${digitPart})[jJ]\\b` }
		]
	};
	const commentType = {
		begin: lookahead(/# type:/),
		className: "comment",
		contains: [{ begin: /# type:/ }, {
			begin: /#/,
			end: /\b\B/,
			endsWithParent: true
		}],
		end: /$/,
		keywords
	};
	const params = {
		className: "params",
		variants: [{
			begin: /\(\s*\)/,
			className: "",
			skip: true
		}, {
			begin: /\(/,
			contains: [
				"self",
				prompt,
				number,
				string,
				hljs.HASH_COMMENT_MODE
			],
			end: /\)/,
			excludeBegin: true,
			excludeEnd: true,
			keywords
		}]
	};
	subst.contains = [
		string,
		number,
		prompt
	];
	return {
		aliases: [...STABLE_PYTHON_ALIASES],
		contains: [
			prompt,
			number,
			{ begin: /\bself\b/ },
			{
				beginKeywords: "if",
				relevance: 0
			},
			string,
			commentType,
			hljs.HASH_COMMENT_MODE,
			{
				contains: [
					hljs.UNDERSCORE_TITLE_MODE,
					params,
					{
						begin: /->/,
						endsWithParent: true,
						keywords
					}
				],
				end: /:/,
				illegal: /[${=;\n,]/,
				variants: [{
					beginKeywords: "def",
					className: "function"
				}, {
					beginKeywords: "class",
					className: "class"
				}]
			},
			{
				begin: /^[\t ]*@/,
				className: "meta",
				contains: [
					number,
					params,
					string
				],
				end: /(?=#)|$/
			}
		],
		illegal: /(<\/|->|\?)|=>/,
		keywords,
		name: "Python"
	};
}
const ensureStablePythonGrammar = (lowlight, language) => {
	if (language !== "python" || !lowlight || patchedLowlights.has(lowlight)) return;
	if (typeof lowlight.register !== "function") return;
	lowlight.register("python", pythonBrowserSafe);
	if (typeof lowlight.registerAlias === "function") lowlight.registerAlias("python", [...STABLE_PYTHON_ALIASES]);
	patchedLowlights.add(lowlight);
};

//#endregion
//#region src/lib/setCodeBlockToDecorations.ts
const CODE_LINE_TO_DECORATIONS = /* @__PURE__ */ new WeakMap();
function getHighlightNodes(result) {
	return result.value || result.children || [];
}
function parseNodes(nodes, className = []) {
	return nodes.flatMap((node) => {
		const classes = [...className, ...node.properties ? node.properties.className : []];
		if (node.children) return parseNodes(node.children, classes);
		return {
			classes,
			text: node.value
		};
	});
}
function normalizeTokens(tokens) {
	const lines = [[]];
	let currentLine = lines[0];
	for (const token of tokens) {
		const tokenLines = token.text.split("\n");
		for (let i = 0; i < tokenLines.length; i++) {
			const content = tokenLines[i];
			if (content) currentLine.push({
				classes: token.classes,
				content
			});
			if (i < tokenLines.length - 1) {
				lines.push([]);
				currentLine = lines.at(-1);
			}
		}
	}
	return lines;
}
function codeBlockToDecorations(editor, [block, blockPath]) {
	const { defaultLanguage, ...options } = editor.getOptions(BaseCodeBlockPlugin);
	const lowlight = options.lowlight;
	const text = block.children.map((line) => NodeApi.string(line)).join("\n");
	const effectiveLanguage = block.lang || defaultLanguage;
	ensureStablePythonGrammar(lowlight, effectiveLanguage);
	let highlighted;
	try {
		if (!effectiveLanguage || effectiveLanguage === "plaintext") highlighted = { value: [] };
		else if (effectiveLanguage === "auto") highlighted = lowlight.highlightAuto(text);
		else highlighted = lowlight.highlight(effectiveLanguage, text);
	} catch (error) {
		const availableLanguages = lowlight.listLanguages();
		if (effectiveLanguage && availableLanguages.includes(effectiveLanguage)) {
			editor.api.debug.warn(`Could not highlight with Highlight.js for language "${effectiveLanguage}". Falling back to plaintext`, "CODE_HIGHLIGHT", error);
			highlighted = { value: [] };
		} else {
			editor.api.debug.warn(`Language "${effectiveLanguage}" is not registered. Falling back to plaintext`);
			highlighted = { value: [] };
		}
	}
	const normalizedTokens = normalizeTokens(parseNodes(getHighlightNodes(highlighted)));
	const blockChildren = block.children;
	const nodeToDecorations = /* @__PURE__ */ new Map();
	const numLines = Math.min(normalizedTokens.length, blockChildren.length);
	for (let index = 0; index < numLines; index++) {
		const lineTokens = normalizedTokens[index];
		const element = blockChildren[index];
		if (!nodeToDecorations.has(element)) nodeToDecorations.set(element, []);
		let start = 0;
		for (const token of lineTokens) {
			const length = token.content.length;
			if (!length) continue;
			const end = start + length;
			const decoration = {
				anchor: {
					offset: start,
					path: [
						...blockPath,
						index,
						0
					]
				},
				className: token.classes.join(" "),
				focus: {
					offset: end,
					path: [
						...blockPath,
						index,
						0
					]
				},
				[KEYS.codeSyntax]: true
			};
			nodeToDecorations.get(element).push(decoration);
			start = end;
		}
	}
	return nodeToDecorations;
}
function setCodeBlockToDecorations(editor, [block, blockPath]) {
	const decorations = codeBlockToDecorations(editor, [block, blockPath]);
	for (const [node, decs] of decorations.entries()) CODE_LINE_TO_DECORATIONS.set(node, decs);
}
function resetCodeBlockDecorations(codeBlock) {
	codeBlock.children.forEach((line) => {
		CODE_LINE_TO_DECORATIONS.delete(line);
	});
}

//#endregion
//#region src/lib/transforms/deleteStartSpace.ts
const whitespaceRegex = /\s/;
/** If there is a whitespace character at the start of the code line, delete it. */
const deleteStartSpace = (editor, { codeLine }) => {
	const [, codeLinePath] = codeLine;
	const codeLineStart = editor.api.start(codeLinePath);
	const codeLineEnd = codeLineStart && editor.api.after(codeLineStart);
	const spaceRange = codeLineEnd && editor.api.range(codeLineStart, codeLineEnd);
	const spaceText = editor.api.string(spaceRange);
	if (whitespaceRegex.test(spaceText)) {
		editor.tf.delete({ at: spaceRange });
		return true;
	}
	return false;
};

//#endregion
//#region src/lib/transforms/indentCodeLine.ts
const nonWhitespaceRegex = /\S/;
/**
* Indent if:
*
* - The selection is expanded OR
* - There are no non-whitespace characters left of the cursor Indentation = 2
*   spaces.
*/
const indentCodeLine = (editor, { codeLine, indentDepth = 2 }) => {
	const [, codeLinePath] = codeLine;
	const codeLineStart = editor.api.start(codeLinePath);
	const indent = " ".repeat(indentDepth);
	if (!editor.api.isExpanded()) {
		const cursor = editor.selection?.anchor;
		const range = editor.api.range(codeLineStart, cursor);
		const text = editor.api.string(range);
		if (nonWhitespaceRegex.test(text)) {
			editor.tf.insertText(indent, { at: editor.selection });
			return;
		}
	}
	editor.tf.insertText(indent, { at: codeLineStart });
};

//#endregion
//#region src/lib/transforms/outdentCodeLine.ts
/** Outdent the code line. Remove 2 whitespace characters if any. */
const outdentCodeLine = (editor, { codeBlock, codeLine }) => {
	if (deleteStartSpace(editor, {
		codeBlock,
		codeLine
	})) deleteStartSpace(editor, {
		codeBlock,
		codeLine
	});
};

//#endregion
//#region src/lib/transforms/unwrapCodeBlock.ts
const unwrapCodeBlock = (editor) => {
	if (!editor.selection) return;
	const codeBlockType = editor.getType(KEYS.codeBlock);
	const defaultType = editor.getType(KEYS.p);
	editor.tf.withoutNormalizing(() => {
		const codeBlockEntries = editor.api.nodes({
			at: editor.selection,
			match: { type: codeBlockType }
		});
		const reversedCodeBlockEntries = Array.from(codeBlockEntries).reverse();
		for (const codeBlockEntry of reversedCodeBlockEntries) {
			const codeLineEntries = NodeApi.children(editor, codeBlockEntry[1]);
			for (const [, path] of codeLineEntries) editor.tf.setNodes({ type: defaultType }, { at: path });
			editor.tf.unwrapNodes({
				at: codeBlockEntry[1],
				match: { type: codeBlockType },
				split: true
			});
		}
	});
};

//#endregion
//#region src/lib/withInsertDataCodeBlock.ts
const withInsertDataCodeBlock = ({ editor, tf: { insertData }, type: codeBlockType }) => ({ transforms: { insertData(data) {
	const text = data.getData("text/plain");
	const vscodeDataString = data.getData("vscode-editor-data");
	const codeLineType = editor.getType(KEYS.codeLine);
	if (vscodeDataString) try {
		const vscodeData = JSON.parse(vscodeDataString);
		const lines = text.split("\n");
		const [blockAbove$1] = editor.api.block() ?? [];
		if (blockAbove$1 && [codeBlockType, codeLineType].includes(blockAbove$1?.type)) {
			if (lines[0]) editor.tf.insertText(lines[0]);
			if (lines.length > 1) {
				const nodes = lines.slice(1).map((line) => ({
					children: [{ text: line }],
					type: codeLineType
				}));
				editor.tf.insertNodes(nodes);
			}
		} else {
			const node = {
				children: lines.map((line) => ({
					children: [{ text: line }],
					type: codeLineType
				})),
				lang: vscodeData?.mode,
				type: codeBlockType
			};
			editor.tf.insertNodes(node, { select: true });
		}
		return;
	} catch (_error) {}
	const [blockAbove] = editor.api.block() ?? [];
	if (blockAbove && [codeBlockType, codeLineType].includes(blockAbove?.type) && text?.includes("\n")) {
		const lines = text.split("\n");
		if (lines[0]) editor.tf.insertText(lines[0]);
		if (lines.length > 1) {
			const nodes = lines.slice(1).map((line) => ({
				children: [{ text: line }],
				type: codeLineType
			}));
			editor.tf.insertNodes(nodes);
		}
		return;
	}
	insertData(data);
} } });

//#endregion
//#region src/lib/withInsertFragmentCodeBlock.ts
function extractCodeLinesFromCodeBlock(node) {
	return node.children;
}
const withInsertFragmentCodeBlock = ({ editor, tf: { insertFragment }, type: codeBlockType }) => ({ transforms: { insertFragment(fragment) {
	const [blockAbove] = editor.api.block() ?? [];
	const codeLineType = editor.getType(KEYS.codeLine);
	function convertNodeToCodeLine(node) {
		return {
			children: [{ text: NodeApi.string(node) }],
			type: codeLineType
		};
	}
	if (blockAbove && [codeBlockType, codeLineType].includes(blockAbove?.type)) return insertFragment(fragment.flatMap((node) => {
		const element = node;
		return element.type === codeBlockType ? extractCodeLinesFromCodeBlock(element) : convertNodeToCodeLine(element);
	}));
	return insertFragment(fragment);
} } });

//#endregion
//#region src/lib/withNormalizeCodeBlock.tsx
/** Normalize code block node to force the pre>code>div.codeline structure. */
const withNormalizeCodeBlock = ({ editor, getOptions, tf: { normalizeNode }, type }) => ({ transforms: { normalizeNode([node, path]) {
	if (node.type === type && getOptions().lowlight) setCodeBlockToDecorations(editor, [node, path]);
	normalizeNode([node, path]);
	if (!ElementApi.isElement(node)) return;
	const codeBlockType = editor.getType(KEYS.codeBlock);
	const codeLineType = editor.getType(KEYS.codeLine);
	if (node.type === codeBlockType) {
		const nonCodeLine = Array.from(NodeApi.children(editor, path)).find(([child]) => child.type !== codeLineType);
		if (nonCodeLine) editor.tf.setNodes({ type: codeLineType }, { at: nonCodeLine[1] });
	}
} } });

//#endregion
//#region src/lib/withCodeBlock.ts
const withCodeBlock = (ctx) => {
	const { editor, getOptions, tf: { apply, deleteBackward, insertBreak, resetBlock, selectAll, tab }, type } = ctx;
	return { transforms: {
		apply(operation) {
			let shouldRedecorate = false;
			if (getOptions().lowlight && operation.type === "set_node") {
				const entry = editor.api.node(operation.path);
				const touchesLang = "lang" in (operation.properties ?? {}) || "lang" in (operation.newProperties ?? {});
				const langChanged = operation.properties?.lang !== operation.newProperties?.lang;
				if (entry?.[0].type === type && touchesLang && langChanged) {
					resetCodeBlockDecorations(entry[0]);
					shouldRedecorate = true;
				}
			}
			apply(operation);
			if (shouldRedecorate) editor.api.redecorate();
		},
		insertBreak() {
			const apply$1 = () => {
				if (!editor.selection) return;
				const res = getCodeLineEntry(editor, {});
				if (!res) return;
				const { codeBlock, codeLine } = res;
				const indentDepth = getIndentDepth(editor, {
					codeBlock,
					codeLine
				});
				insertBreak();
				indentCodeLine(editor, {
					codeBlock,
					codeLine,
					indentDepth
				});
				return true;
			};
			if (apply$1()) return;
			insertBreak();
		},
		deleteBackward(unit) {
			const apply$1 = () => {
				if (!editor.selection || editor.api.isExpanded()) return;
				const res = getCodeLineEntry(editor, {});
				if (!res) return;
				const { codeLine } = res;
				const [, codeLinePath] = codeLine;
				if (!editor.api.isStart(editor.selection.anchor, codeLinePath)) return;
				const previousCodeLine = editor.api.previous({
					at: codeLinePath,
					match: { type: editor.getType(KEYS.codeLine) }
				});
				const codeLineText = editor.api.string(codeLinePath);
				if (!previousCodeLine) {
					if (codeLineText.length > 0) return true;
					return;
				}
				if (codeLineText.length > 0) return;
				const previousLineEnd = editor.api.end(previousCodeLine[1]);
				editor.tf.removeNodes({ at: codeLinePath });
				if (previousLineEnd) editor.tf.select(previousLineEnd);
				return true;
			};
			if (apply$1()) return;
			deleteBackward(unit);
		},
		resetBlock(options) {
			if (editor.api.block({
				at: options?.at,
				match: { type }
			})) {
				unwrapCodeBlock(editor);
				return;
			}
			return resetBlock(options);
		},
		selectAll: () => {
			const apply$1 = () => {
				const codeBlock = editor.api.above({ match: { type } });
				if (!codeBlock) return;
				if (editor.api.isAt({ end: true }) && editor.api.isAt({ start: true })) return;
				editor.tf.select(codeBlock[1]);
				return true;
			};
			if (apply$1()) return true;
			return selectAll();
		},
		tab: (options) => {
			const apply$1 = () => {
				const codeLineType = editor.getType("code_line");
				const _codeLines = editor.api.nodes({ match: { type: codeLineType } });
				const codeLines = Array.from(_codeLines);
				if (codeLines.length > 0) {
					const [, firstLinePath] = codeLines[0];
					const codeBlock = editor.api.parent(firstLinePath);
					if (!codeBlock) return;
					editor.tf.withoutNormalizing(() => {
						for (const codeLine of codeLines) if (options.reverse) outdentCodeLine(editor, {
							codeBlock,
							codeLine
						});
						else indentCodeLine(editor, {
							codeBlock,
							codeLine
						});
					});
					return true;
				}
			};
			if (apply$1()) return true;
			return tab(options);
		},
		...withInsertDataCodeBlock(ctx).transforms,
		...withInsertFragmentCodeBlock(ctx).transforms,
		...withNormalizeCodeBlock(ctx).transforms
	} };
};

//#endregion
//#region src/lib/BaseCodeBlockPlugin.ts
const BaseCodeLinePlugin = createTSlatePlugin({
	key: KEYS.codeLine,
	node: {
		isElement: true,
		isStrictSiblings: true
	}
});
const BaseCodeSyntaxPlugin = createSlatePlugin({
	key: KEYS.codeSyntax,
	node: { isLeaf: true }
});
const BaseCodeBlockPlugin = createTSlatePlugin({
	key: KEYS.codeBlock,
	inject: { plugins: { [KEYS.html]: { parser: { query: ({ editor }) => !editor.api.some({ match: { type: editor.getType(KEYS.codeLine) } }) } } } },
	node: { isElement: true },
	options: {
		defaultLanguage: null,
		lowlight: null
	},
	parsers: { html: { deserializer: htmlDeserializerCodeBlock } },
	plugins: [BaseCodeLinePlugin, BaseCodeSyntaxPlugin],
	render: { as: "pre" },
	rules: {
		delete: { empty: "reset" },
		match: ({ editor, rule }) => ["break.empty", "delete.empty"].includes(rule) && isCodeBlockEmpty(editor)
	},
	decorate: ({ editor, entry: [node, path], getOptions, type }) => {
		if (!getOptions().lowlight) return [];
		const codeLineType = editor.getType(KEYS.codeLine);
		if (node.type === type && !CODE_LINE_TO_DECORATIONS.get(node.children[0])) setCodeBlockToDecorations(editor, [node, path]);
		if (node.type === codeLineType) return CODE_LINE_TO_DECORATIONS.get(node) || [];
		return [];
	}
}).overrideEditor(withCodeBlock).extendTransforms(({ editor }) => ({ toggle: () => {
	editor.tf.toggleBlock(editor.getType(KEYS.codeBlock));
} }));

//#endregion
export { isCodeBlockEmpty as _, withNormalizeCodeBlock as a, htmlDeserializerCodeBlock as b, unwrapCodeBlock as c, deleteStartSpace as d, CODE_LINE_TO_DECORATIONS as f, ensureStablePythonGrammar as g, setCodeBlockToDecorations as h, withCodeBlock as i, outdentCodeLine as l, resetCodeBlockDecorations as m, BaseCodeLinePlugin as n, withInsertFragmentCodeBlock as o, codeBlockToDecorations as p, BaseCodeSyntaxPlugin as r, withInsertDataCodeBlock as s, BaseCodeBlockPlugin as t, indentCodeLine as u, getIndentDepth as v, getCodeLineEntry as y };
