// @platejs/code-block@53.0.0/dist/index.js
import { _ as isCodeBlockEmpty, a as withNormalizeCodeBlock, b as htmlDeserializerCodeBlock, c as unwrapCodeBlock, d as deleteStartSpace, f as CODE_LINE_TO_DECORATIONS, g as ensureStablePythonGrammar, h as setCodeBlockToDecorations, i as withCodeBlock, l as outdentCodeLine, m as resetCodeBlockDecorations, n as BaseCodeLinePlugin, o as withInsertFragmentCodeBlock, p as codeBlockToDecorations, r as BaseCodeSyntaxPlugin, s as withInsertDataCodeBlock, t as BaseCodeBlockPlugin, u as indentCodeLine, v as getIndentDepth, y as getCodeLineEntry } from "./BaseCodeBlockPlugin-stnOioFF.js";
import { KEYS, createRuleFactory } from "platejs";

//#region src/lib/queries/isSelectionAtCodeBlockStart.ts
/** Is the selection at the start of the first code line in a code block */
const isSelectionAtCodeBlockStart = (editor) => {
	const { selection } = editor;
	if (!selection || editor.api.isExpanded()) return false;
	const { codeBlock } = getCodeLineEntry(editor) ?? {};
	if (!codeBlock) return false;
	return editor.api.isStart(selection.anchor, codeBlock[1]);
};

//#endregion
//#region src/lib/transforms/insertCodeBlock.ts
/**
* Insert a code block: set the node to code line and wrap it with a code block.
* If the cursor is not at the block start, insert break before.
*/
const insertCodeBlock = (editor, insertNodesOptions = {}) => {
	if (!editor.selection || editor.api.isExpanded()) return;
	const matchCodeElements = (node) => node.type === KEYS.codeBlock || node.type === KEYS.codeLine;
	if (editor.api.some({ match: matchCodeElements })) return;
	if (!editor.api.isAt({ start: true })) editor.tf.insertBreak();
	editor.tf.setNodes({
		children: [{ text: "" }],
		type: KEYS.codeLine
	}, insertNodesOptions);
	editor.tf.wrapNodes({
		children: [],
		type: KEYS.codeBlock
	}, insertNodesOptions);
};

//#endregion
//#region src/lib/transforms/insertCodeLine.ts
/** Insert a code line starting with indentation. */
const insertCodeLine = (editor, indentDepth = 0) => {
	if (editor.selection) {
		const indent = " ".repeat(indentDepth);
		editor.tf.insertNodes({
			children: [{ text: indent }],
			type: editor.getType(KEYS.codeLine)
		});
	}
};

//#endregion
//#region src/lib/transforms/insertEmptyCodeBlock.ts
/**
* Called by toolbars to make sure a code-block gets inserted below a paragraph
* rather than awkwardly splitting the current selection.
*/
const insertEmptyCodeBlock = (editor, { defaultType = editor.getType(KEYS.p), insertNodesOptions } = {}) => {
	if (!editor.selection) return;
	if (editor.api.isExpanded() || !editor.api.isEmpty(editor.selection, { block: true })) editor.tf.insertNodes(editor.api.create.block({
		children: [{ text: "" }],
		type: defaultType
	}), {
		nextBlock: true,
		select: true,
		...insertNodesOptions
	});
	insertCodeBlock(editor, insertNodesOptions);
};

//#endregion
//#region src/lib/transforms/setCodeBlockContent.ts
const setCodeBlockContent = (editor, { code, element }) => {
	editor.tf.replaceNodes(code.split("\n").map((line) => ({
		children: [{ text: line }],
		type: editor.getType(KEYS.codeLine)
	})), {
		at: element,
		children: true
	});
	editor.api.redecorate();
};

//#endregion
//#region src/lib/transforms/toggleCodeBlock.ts
const toggleCodeBlock = (editor) => {
	if (!editor.selection) return;
	const codeBlockType = editor.getType(KEYS.codeBlock);
	const codeLineType = editor.getType(KEYS.codeLine);
	const isActive = editor.api.some({ match: { type: codeBlockType } });
	editor.tf.withoutNormalizing(() => {
		unwrapCodeBlock(editor);
		if (!isActive) {
			editor.tf.setNodes({ type: codeLineType });
			const codeBlock = {
				children: [],
				type: codeBlockType
			};
			editor.tf.wrapNodes(codeBlock);
		}
	});
};

//#endregion
//#region src/lib/CodeBlockRules.ts
const isCodeBlockInputBlocked = (editor) => editor.api.some({ match: { type: [editor.getType(KEYS.codeBlock)] } });
const insertCodeBlockAtPath = (editor, path) => {
	editor.tf.removeNodes({ at: path });
	editor.tf.insertNodes({
		children: [{
			children: [{ text: "" }],
			type: editor.getType(KEYS.codeLine)
		}],
		type: editor.getType(KEYS.codeBlock)
	}, { at: path });
	const start = editor.api.start([...path, 0]);
	if (start) editor.tf.select(start);
};
const CodeBlockRules = { markdown: createRuleFactory({
	type: "blockFence",
	fence: "```",
	block: KEYS.p,
	enabled: ({ editor }) => !isCodeBlockInputBlocked(editor),
	priority: 100,
	apply: ({ editor, on }, match) => {
		if (on === "break") insertCodeBlockAtPath(editor, match.path);
		else insertCodeBlockAtPath(editor, match.path);
		return true;
	}
}) };

//#endregion
//#region src/lib/formatter/jsonFormatter.ts
const formatJson = (code) => {
	try {
		return JSON.stringify(JSON.parse(code), null, 2);
	} catch (_error) {
		return code;
	}
};
const isValidJson = (code) => {
	try {
		JSON.parse(code);
		return true;
	} catch (_error) {
		return false;
	}
};

//#endregion
//#region src/lib/formatter/formatter.ts
const supportedLanguages = new Set(["json"]);
const isLangSupported = (lang) => Boolean(lang && supportedLanguages.has(lang));
const isValidSyntax = (code, lang) => {
	if (!isLangSupported(lang)) return false;
	switch (lang) {
		case "json": return isValidJson(code);
		default: return false;
	}
};
const formatCodeBlock = (editor, { element }) => {
	const { lang } = element;
	if (!lang || !isLangSupported(lang)) return;
	const code = editor.api.string(element);
	if (isValidSyntax(code, lang)) setCodeBlockContent(editor, {
		code: formatCode(code, lang),
		element
	});
};
const formatCode = (code, lang) => {
	switch (lang) {
		case "json": return formatJson(code);
		default: return code;
	}
};

//#endregion
export { BaseCodeBlockPlugin, BaseCodeLinePlugin, BaseCodeSyntaxPlugin, CODE_LINE_TO_DECORATIONS, CodeBlockRules, codeBlockToDecorations, deleteStartSpace, ensureStablePythonGrammar, formatCodeBlock, formatJson, getCodeLineEntry, getIndentDepth, htmlDeserializerCodeBlock, indentCodeLine, insertCodeBlock, insertCodeLine, insertEmptyCodeBlock, isCodeBlockEmpty, isLangSupported, isSelectionAtCodeBlockStart, isValidJson, isValidSyntax, outdentCodeLine, resetCodeBlockDecorations, setCodeBlockContent, setCodeBlockToDecorations, toggleCodeBlock, unwrapCodeBlock, withCodeBlock, withInsertDataCodeBlock, withInsertFragmentCodeBlock, withNormalizeCodeBlock };
